### PR TITLE
[dyno] Add support for associating additional locations with AST

### PR DESCRIPTION
**Pull request by dlongnecke-cray**
This PR allows developers to associate more than one location with an AST node. The mechanism is built with macros in order to make it easy to add more locations as needed.

For example: when parsing a dot expression `x.foo`, an `Identifier` is created for `x` and a `Dot` is created for `x.foo` (with `x` as a child).

The location normally created for the `Dot` will span the entire text `x.foo`. However, it is impossible to tell precisely where `foo` is in `x.    foo`.

This PR adds support for associating the location of a "dot field" with a `Dot` AST by calling `Builder::noteDotFieldLocation`. As well, it adds support for pairing declaration names with `NamedDecl` using `Builder::noteDeclNameLocation`.

Additional locations can be added to "chpl/uast/all-location-maps.h".

This support is needed in order to improve the accuracy of the language server's "go to declaration" feature. It may enable more precise tooling in general.

While here, restructure the parser grammar file a little bit by defining a new struct type `TextLocation` as `YYLTYPE` and moving it into its own header file. This allows locations to be stored inside of the types such as `TypeDeclParts` that are constructed during parsing.

Move the handlers for `begin` and `with` into builders in the `ParserContext` to clean up the grammar file. Also add the builder `buildDot` for dot expressions.

Reviewed by @DanilaFe. Thanks!

TESTING

- [x] `linux64`, `standard`

### CODE REVIEW

File: frontend/include/chpl/framework/ID.h [hunk 1]
```diff
@@ -217,7 +217,7 @@ class ID final {
     return symbolPath_.isEmpty();
```
Comment (DanilaFe) [inline frontend/include/chpl/framework/ID.h:217]: Not really part of the code diff, but you should stop 80col-wrapping your PR description in GitHub. It's harder to read on GitHub itself because it wraps, and GitHub can auto-wrap it to 80 columns when putting into the merge commit's text.

File: frontend/lib/parsing/parsing-queries.cpp [hunk 2]
```diff
@@ -388,21 +388,26 @@ void countTokens(Context* context, UniqueString path, ParserStats* parseStats) {
   }
 }
 
+static const BuilderResult*
+builderResultPtr(Context* context, ID id, UniqueString& pathOut) {
```
Comment (DanilaFe) [inline frontend/lib/parsing/parsing-queries.cpp:392]: I think the Dyno convention has been to name these functions with 'OrNull'. E.g.:

```
  const ResolvedExpression* byAstOrNull(const uast::AstNode* ast) const {
```

So then:
```suggestion
builderResultPtrOrNull(Context* context, ID id, UniqueString& pathOut) {
```

File: frontend/lib/parsing/parsing-queries.cpp [hunk 3]
```diff
@@ -388,21 +388,26 @@ void countTokens(Context* context, UniqueString path, ParserStats* parseStats) {
...
+  }
+  return ret;
+}
+
 const Location& locateId(Context* context, ID id) {
   QUERY_BEGIN(locateId, context, id);
-
   Location result;
 
-  // Ask the context for the filename from the ID
```
Comment (DanilaFe) [inline frontend/lib/parsing/parsing-queries.cpp:396]: You dropped this comment, but I'd move it into `builderResultPtr`.

File: frontend/lib/parsing/parsing-queries.cpp [hunk 4]
```diff
@@ -388,21 +388,26 @@ void countTokens(Context* context, UniqueString path, ParserStats* parseStats) {
...
   UniqueString path;
-  UniqueString parentSymbolPath;
-
-  bool found = context->filePathForId(id, path, parentSymbolPath);
-  if (found) {
-    // Get the result of parsing
```
Comment (DanilaFe) [inline frontend/lib/parsing/parsing-queries.cpp:402]: Ditto with this comment.

File: frontend/lib/parsing/parsing-queries.cpp [hunk 5]
```diff
@@ -411,9 +416,38 @@ const Location& locateId(Context* context, ID id) {
 // this is just a convenient wrapper around locating with the id
 const Location& locateAst(Context* context, const AstNode* ast) {
   CHPL_ASSERT(!ast->isComment() && "cant locate comment like this");
+  CHPL_ASSERT(ast); // TODO: Will crash instead of return nullptr...
```
Comment (DanilaFe) [inline frontend/lib/parsing/parsing-queries.cpp:419]: Pointless assert because the previous line (`ast->isComment()`) would segfault before this assertion would fail. Consider swapping / moving to first line of `locateAst`.

Also, I'm not making much sense of the comment.

File: frontend/include/chpl/uast/BuilderResult.h [hunk 6]
```diff
@@ -34,6 +34,24 @@
 
 #include "llvm/ADT/DenseMap.h"
 
+#define CHPL_ID_LOC_MAP_INNER(ast__, location__) \
+  id##ast__##To##location__##Location_
```
Comment (DanilaFe) [inline frontend/include/chpl/uast/BuilderResult.h:38]: Why do you need the 'inner' macro here?

File: frontend/lib/uast/Builder.cpp [hunk 7]
```diff
@@ -125,6 +125,21 @@ void Builder::noteLocation(AstNode* ast, Location loc) {
   notedLocations_[ast] = loc;
 }
 
+void Builder::noteAdditionalLocation(AstLocMap& m, AstNode* ast,
+                                     Location loc) {
+  if (!ast || loc.isEmpty()) return;
+  CHPL_ASSERT(m.find(ast) == m.end());
+  m.emplace(ast, loc);
```
Comment (DanilaFe) [inline frontend/lib/uast/Builder.cpp:132]: can `loc` be `std::moved`'d here?

File: frontend/lib/uast/Builder.cpp [hunk 8]
```diff
@@ -413,6 +443,22 @@ void Builder::doAssignIDs(AstNode* ast, UniqueString symbolPath, int& i,
...
+    // Also map additional locations to ID.
+    #define CHPL_LOCATION_MAP(ast__, location__) \
+      if (auto x = ast->to##ast__()) { \
+        auto& m1 = CHPL_AST_LOC_MAP(ast__, location__); \
+        auto it = m1.find(x); \
+        if (it != m1.end()) { \
+          auto& m2 = CHPL_ID_LOC_MAP(ast__, location__); \
+          auto& id = x->id(); \
+          CHPL_ASSERT(m2.find(id) == m2.end()); \
+          m2[id] = it->second; \
```
Comment (DanilaFe) [inline frontend/lib/uast/Builder.cpp:456]: You're performing a search twice here. If you preserve the iterator for `m2.find(id)`, you can `emplace_hint` to make the insertion not need to perform a search.

https://en.cppreference.com/w/cpp/container/map/emplace_hint

Alternatively, consider matching the code for `idToLocation` and not bothering with the assertion, or an `id` variable.

File: frontend/lib/uast/BuilderResult.cpp [hunk 9]
```diff
@@ -190,6 +217,18 @@ ID BuilderResult::idToParentId(ID id) const {
...
 }
 
+// Add getters for additional locations that go from AST or ID to location.
+#define CHPL_LOCATION_MAP(ast__, location__) \
+  Location BuilderResult:: \
+  idTo##location__##Location(ID id, UniqueString path) const { \
+    if (!id) return Location(path); \
+    auto& m = CHPL_ID_LOC_MAP(ast__, location__); \
+    auto it = m.find(id); \
+    return (it != m.end()) ? it->second : Location(path); \
```
Comment (DanilaFe) [inline frontend/lib/uast/BuilderResult.cpp:227]: Does this match the existing behavior of Builder for non-special locations?

File: frontend/lib/uast/BuilderResult.cpp [hunk 10]
```diff
@@ -338,6 +377,16 @@ bool BuilderResult::equals(const BuilderResult& other) const {
     }
   }
 
+  // Check that all the additional location maps match.
+  #define CHPL_LOCATION_MAP(ast__, location__) { \
+      auto& m1 = CHPL_ID_LOC_MAP(ast__, location__); \
+      auto& m2 = other.CHPL_ID_LOC_MAP(ast__, location__); \
+      if (m1.size() != m2.size()) return false; \
```
Comment (DanilaFe) [inline frontend/lib/uast/BuilderResult.cpp:384]: note that the 'non-specialized' map code doesn't perform this check. Also note that this check is already performed by `operator==` for DenseMap:

```C++
template <typename DerivedT, typename KeyT, typename ValueT, typename KeyInfoT,
          typename BucketT>
bool operator==(
    const DenseMapBase<DerivedT, KeyT, ValueT, KeyInfoT, BucketT> &LHS,
    const DenseMapBase<DerivedT, KeyT, ValueT, KeyInfoT, BucketT> &RHS) {
  if (LHS.size() != RHS.size())
    return false;
```

File: frontend/test/parsing/testParseAdditionalLocations.cpp [hunk 11]
```diff
@@ -0,0 +1,144 @@
...
+  ErrorGuard guard(ctx);
+
+  auto path = TEST_NAME_FROM_FN_NAME(ctx);
+  std::string contents =
+    "x1.field;\n"   // 1:4
+    "x1. field;\n"  // 2:5
+    "x1.\n"         //
+    "   field;\n"   // 4:4
+    "x1\n"          //
+    ".field;\n";    // 6:2
```
Comment (DanilaFe) [inline frontend/test/parsing/testParseAdditionalLocations.cpp:53]: Consider varying the name of the field to catch parser hysteresis.

File: frontend/lib/parsing/parser-yyltype.h
Comment (DanilaFe) [inline frontend/lib/parsing/parser-yyltype.h:1]: Why do you need this file? I don't even see references to `first_line` in the rest of the diff.

File: frontend/lib/parsing/ParserContext.h [hunk 12]
```diff
@@ -274,6 +280,18 @@ struct ParserContext {
...
+                              YYLTYPE locTaskIntentList,
+                              YYLTYPE locRightParen,
+                              ParserExprList* exprList);
+
+  CommentsAndStmt buildBeginStmt(YYLTYPE location, YYLTYPE locBegin,
+                                 YYLTYPE locWithClause,
+                                 YYLTYPE locStmt,
+                                 WithClause* withClause,
+                                 CommentsAndStmt stmt);
+
```
Comment (DanilaFe) [inline frontend/lib/parsing/ParserContext.h:294]: Why did you need to change these for `Dot` fields and `NamedDecl`s?

File: frontend/include/chpl/framework/ID.h (hunk 1 above)
Comment (dlongnecke-cray) [reply frontend/include/chpl/framework/ID.h:217]: Does it really do that now? I guess I could give it a try this PR. But what zoom are you reading GitHub at if it wraps for you?!

File: frontend/lib/parsing/parsing-queries.cpp (hunk 5 above)
Comment (dlongnecke-cray) [reply frontend/lib/parsing/parsing-queries.cpp:419]: Ideally, the query wouldn't crash?

File: frontend/include/chpl/uast/BuilderResult.h (hunk 6 above)
Comment (dlongnecke-cray) [reply frontend/include/chpl/uast/BuilderResult.h:38]: Otherwise the concatenated macro tokens won't expand.

File: frontend/lib/uast/BuilderResult.cpp (hunk 9 above)
Comment (dlongnecke-cray) [reply frontend/lib/uast/BuilderResult.cpp:227]:
```cpp
  /** Find the Location for a particular ID.
      Returns a location just to path if none is found. */
  Location idToLocation(ID id, UniqueString path) const;
```

File: frontend/lib/uast/BuilderResult.cpp (hunk 10 above)
Comment (dlongnecke-cray) [reply frontend/lib/uast/BuilderResult.cpp:384]: That's true, but I didn't want to make assumptions about the map code.

File: frontend/lib/parsing/parser-yyltype.h
Comment (dlongnecke-cray) [reply frontend/lib/parsing/parser-yyltype.h:1]: I'm not sure what you mean, this type is `YYLTYPE` so `first_line` is used everywhere. I pulled it out into its own header so that the rest of the parser types could refer to it at will (they could refer to it before the start of the grammar rules, before). I'll add a comment at the top of the file.

File: frontend/lib/parsing/ParserContext.h (hunk 12 above)
Comment (dlongnecke-cray) [reply frontend/lib/parsing/ParserContext.h:294]: In general, I think it's the right thing to do to pull any parser action more than one call long out into a separate builder. This is something we didn't always do when we rushed to get the parser in, but I think it's the right thing to do now if your work needs to touch a particular multi-line action. I pulled these out because I had to correct them to fix bugs.

File: frontend/include/chpl/uast/BuilderResult.h [hunk 13]
```diff
@@ -34,6 +34,24 @@
...
+
+// These helpers make it easy for the Builder and BuilderResult types to
+// manipulate their internal location maps without having to reconstruct
+// their names each time with concatenation.
+//
+// Use to get the name of an additional <ID, Location> map.
```
[the remark this answers is not in the document]
Comment (dlongnecke-cray) [reply frontend/include/chpl/uast/BuilderResult.h:44]: I deliberately chose not to comment that one because it shouldn't be invoked, I'll just write a little comment on those two as well.

File: frontend/include/chpl/framework/ID.h (hunk 1 above)
Comment (dlongnecke-cray) [reply frontend/include/chpl/framework/ID.h:217]: I'm taking a leap into a brave new world...

File: frontend/lib/parsing/parsing-queries.cpp (hunk 3 above)
Comment (dlongnecke-cray) [reply frontend/lib/parsing/parsing-queries.cpp:396]: I kind of feel like what it's doing is self evident from the name of the call.

File: frontend/lib/uast/BuilderResult.cpp (hunk 10 above)
Comment (DanilaFe) [reply frontend/lib/uast/BuilderResult.cpp]: I think that any reasonable implementation will do that check and it just gets in the way here. But your call whether to keep it.

File: frontend/include/chpl/parsing/parsing-queries.h [hunk 14]
```diff
@@ -179,6 +179,22 @@ const Location& locateId(Context* context, ID id);
...
 const Location& locateAst(Context* context, const uast::AstNode* ast);
 
+/** Also define getters for additional locations such as e.g., dot fields.
+    A complete list can be seen in "chpl/uast/all-location-maps.h".
+    The form is e.g., `locateDotFieldWithId(Context* context, ID id)` or
+    `locateDotFieldWithAst(Context* context, const Dot* dot)`.
+
+    Additional location maps for things like dot fields are required because
+    the parser does not keep track of the location of every piece of text
+    when parsing a file into AST.
```
Comment (DanilaFe) [inline frontend/include/chpl/parsing/parsing-queries.h:189]: And because the field is not an AST node in itself.

File: frontend/lib/parsing/parsing-queries.cpp [hunk 15]
```diff
@@ -388,32 +388,71 @@ void countTokens(Context* context, UniqueString path, ParserStats* parseStats) {
...
 
   return QUERY_END(result);
 }
 
 // this is just a convenient wrapper around locating with the id
 const Location& locateAst(Context* context, const AstNode* ast) {
+  CHPL_ASSERT(ast);
   CHPL_ASSERT(!ast->isComment() && "cant locate comment like this");
+  // TODO: This will crash if 'ast' is null, can we fix that?
   return locateId(context, ast->id());
```
Comment (DanilaFe) [inline frontend/lib/parsing/parsing-queries.cpp:425]: IMO, `null` is just not a valid input to `locateAst`, so there's nothing to be done. Garbge in, garbage out.

File: frontend/lib/parsing/parser-yyltype.h
Comment (DanilaFe) [reply frontend/lib/parsing/parser-yyltype.h:1]: You say 'pulled it out', but from where?

Comment (dlongnecke-cray) [reply frontend/lib/parsing/parser-yyltype.h:1]: It was defined implicitly by `chpl.ypp` before the grammar rules start, but I've defined it explicitly in the header file. You can see the "define YYCHPL_LTYPE" stuff in the deleted diff.

Comment (DanilaFe) [reply frontend/lib/parsing/parser-yyltype.h:1]: Saw your comment. Thank you, that (and your comment here) clarifies it.

**Review by DanilaFe — APPROVED** [review]
